### PR TITLE
[homematic] Fix large number parsing

### LINKED ISSUE #17577: [Homematic Binding 4.2.2] GATEWAY-EXTRAS do not load/refresh if systemvariable has special content
## Expected Behavior
If the binding is scanning (new) things, GATEWAY-EXTRAS should be there no matter what the content of any string-variable is
A refresh should refresh all variables no matter what the content of any string-variable is

## Current Behavior
I do have a fresh installation of Homematic Raspberrymatic 3.77.7 and openhabian 4.2.2.
No additional bridges within HM and in openhabian only homematic binding has been added.
After the initialization/scan the GATEWAY-EXTRAS thing is missing if a following systevariable in HM exists:
Name: doesn't matter
type: string
content: >= 11 numbers

event.log warnings:
2024-10-16 16:48:22.465 [WARN ] [ternal.handler.HomematicThingHandler] - For input string: "12345678901"
2024-10-16 16:50:41.946 [WARN ] [ternal.handler.HomematicThingHandler] - For input string: "1234567890123"
2024-10-16 16:54:20.773 [WARN ] [ternal.handler.HomematicThingHandler] - For input string: "-12345678901"
2024-10-16 16:55:13.529 [WARN ] [ternal.handler.HomematicThingHandler] - For input string: "+12345678901"

-> although it is only a warning no variables are refreshed!
-> If the variable exists before the binding ist installed, GATEWAY-EXTRAS thing will not be found!

If the variable contains a letter everything works fine. If it is shorter and contains only number it's also ok.
E. g. works with:
-1234567890x
+1234567890x
x1234567890x
111
12345678901xyz

## Possible Solution
change the content of the variable :-)

## Steps to Reproduce (for Bugs)
Easy! See description above. Just add a string variable wir 11 digits and see wether GATEWAY-EXTRAS can be refreshed!

## Your Environment
I do have a fresh installation of Homematic Raspberrymatic 3.77.7 on Raspi 3B+ and openhabian 4.2.2 on Raspi 4B

### PR DESCRIPTION
**Pull request by lsiepel**
- Fixes: https://github.com/openhab/openhab-addons/issues/17577 
 
Occurs when a number is overflowing. This can be observed from the logs
 
```
2024-10-16 16:48:22.465 [WARN ] [ternal.handler.HomematicThingHandler] - For input string: "12345678901"
2024-10-16 16:50:41.946 [WARN ] [ternal.handler.HomematicThingHandler] - For input string: "1234567890123"
2024-10-16 16:54:20.773 [WARN ] [ternal.handler.HomematicThingHandler] - For input string: "-12345678901"
2024-10-16 16:55:13.529 [WARN ] [ternal.handler.HomematicThingHandler] - For input string: "+12345678901"
```

Jar for testing openHAB 5.2.0 or later: https://1drv.ms/u/c/70ea7ac46bc61c73/IQDkQDQrnOtnRauFlBUXP--pASFBeGdd2kXgu-PVKB-21Ug?e=fQJHy4

Candidate for backporting

### CODE REVIEW

**Comment by maniac103** [comment]
Just so that I understand correctly: the actual fix here is catching the exception, not the overflow, right? That is, before it threw the exception and nothing handled it, now the exception is handled and `null` is returned ... is that correct? If so, I wonder whether a more correct/complete fix wouldn't be changing the `convertToType(String)` method to use the same methods (`toNumber()`, `toString()` etc.) as `convertToType(HmDatapoint, String` already does?

**Comment by lsiepel** [comment]
It is now parsed as string, not as number. If that is a requirement, we need BigDecimal (everywhere).
I also wonder what uses such a huge number

**Comment by maniac103** [comment]
> It is now parsed as string, not as number. If that is a requirement, we need BigDecimal (everywhere).
> I also wonder what uses such a huge number

For the case of the gateway variable that totally makes sense (which is why I think the change to use the DP value type makes sense). The number probably actually is a phone number or something.
My comment was more targeted towards having two `convertToType()` methods were one throws on encountering invalid values and the other one returns null. The latter seems more safe behavior to me for the cases where we don't have a DP when parsing (e.g. in `CcuParamsetDescriptionParser`). Alternatively, we could
- change `convertToType(HmDatapoint, String)` to `convertToType(HmValueType, String)`, changing the current if-else to a `switch` statement (all `HmDataPoint.isFooType()` methods just compare the value type)
- drop `convertToType(String)`
- pass `dp.getType()` instead of `dp` to adapt to the signature change
- make the few callers of `convertToType(String)` call `convertToType(HmValueType, String)` as all of them have the value type handily available

Doing so would have the advantage of allowing to call the `HmDatapoint` constructor with the correct value again.

**Comment by lsiepel** [comment]
Please check again, i consolidated the methods, i think this is much better.
Also updated the jar

**Review by maniac103 — APPROVED** [review]
Yes, this looks much better 👍

**Comment by wborn** [comment]
I want to pass on some remarks from my Codex AI friend 😉:

The overall approach looks good: converting values based on the declared "HmValueType" fixes the reported overflow issue more reliably than inferring the type from the string content.

I found two potential regressions that should be addressed before merging:

1. In "CcuParamsetDescriptionParser", enum values are converted to integers before "assembleDatapoint()" handles the HmIP enum option-name mapping. Symbolic values such as ""AUTO"" can therefore become "null" and ultimately option index "-1". The raw values should be passed to "assembleDatapoint()", which already performs the appropriate conversion.

2. "STRING" values now go through "toString()", which trims whitespace. This changes the contents of string system variables with intentional leading or trailing spaces. String conversion should preserve the original value while still handling blank values as desired.

It would also be useful to add regression tests for an HmIP enum with symbolic values and a string variable containing leading or trailing whitespace.

File: bundles/org.openhab.binding.homematic/src/test/java/org/openhab/binding/homematic/internal/communicator/parser/CcuParamsetDescriptionParserTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import org.openhab.binding.homematic.internal.model.HmDatapoint;
+import org.openhab.binding.homematic.internal.model.HmDevice;
+import org.openhab.binding.homematic.internal.model.HmInterface;
+import org.openhab.binding.homematic.internal.model.HmParamsetType;
+import org.openhab.binding.homematic.internal.model.HmValueType;
+import org.openhab.binding.homematic.internal.model.TclScriptDataEntry;
+import org.openhab.binding.homematic.internal.model.TclScriptDataList;
+
+/**
+ * @author Leo Siepel - Regression test for symbolic HmIP enums
```
Comment (wborn) [inline bundles/org.openhab.binding.homematic/src/test/java/org/openhab/binding/homematic/internal/communicator/parser/CcuParamsetDescriptionParserTest.java]: The static-analysis report flags this new test class because the first `@author` contribution description for a new Java source file must be `Initial contribution`. The same finding is present in [`CcuVariablesAndScriptsParserTest.java`](https://github.com/openhab/openhab-addons/blob/d57dd0f299fc064c929aa337c0d54712817fe053/bundles/org.openhab.binding.homematic/src/test/java/org/openhab/binding/homematic/internal/communicator/parser/CcuVariablesAndScriptsParserTest.java#L38).

Please change both new test classes to:

`@author Leo Siepel - Initial contribution`

**Review by wborn — COMMENTED** [review]
This PR is being reviewed with AI first; a human maintainer review is still required.

The current implementation looks good overall. Converting values according to their declared `HmValueType` fixes the reported issue with numeric-looking string variables, and the earlier HmIP enum and string-whitespace regressions are now handled and covered by regression tests.

The CI build passes, including all Homematic tests.

One minor repository-style issue remains in the `@author` tags of the two new test classes; see the inline comment.

File: bundles/org.openhab.binding.homematic/src/test/java/org/openhab/binding/homematic/internal/communicator/parser/CcuParamsetDescriptionParserTest.java (hunk 1 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.homematic/src/test/java/org/openhab/binding/homematic/internal/communicator/parser/CcuParamsetDescriptionParserTest.java]: :-/ fixed

**Comment by wborn** [comment]
AI found no further issues after re-reviewing the current PR head. The previous `@author` finding has been addressed, and CI passes on the updated commit.

A human maintainer review is still needed.

**Review by maniac103 — APPROVED** [review]

**Comment by lsiepel** [comment]
@openhab/add-ons-maintainers please merge so this can be in M1

**Comment by lolodomo** [comment]
Is it something to backport to branch 5.2.x without any risk ?

**Comment by maniac103** [comment]
I'd say it's pretty low risk, but not zero risk.
